addons/web/static/src/js/components/dropdown_menu.js

```diff
@@ -61,6 +62,24 @@ odoo.define('web.DropdownMenu', function (require) {
             return this.env.device.isMobile;
         }
 
+        /**
+         * By default on mobile, we use the half of the screen as a threshold value to choice the display direction
+         * of the open dropdown
+         *
+         * @return string
+         */
+        get dropdownMenuAlignClass() {
+            if (this.env.device.isMobile) {
+                const threshold = screen.width / 2;
+                const { left, right } = this.el.getBoundingClientRect();
+                if (_t.database.parameters.direction === 'rtl') {
+                    return right > threshold ? 'dropdown-menu-left' : 'dropdown-menu-right';
+                }
+                return left > threshold ? 'dropdown-menu-right' : 'dropdown-menu-left';
+            }
+            return '';
+        }
```

Comment:
This fix should be more generic (both desktop and mobile) and apply similar logic as Popper.js (cf. toggle the `dropdown-menu-right` class only when overflowing)